src/core/lv2/Lv2UridMap.cpp

```diff
@@ -0,0 +1,93 @@
+/*
+ * Lv2UridMap.cpp - Lv2UridMap implementation
+ *
+ * Copyright (c) 2019 Johannes Lorenz <j.git$$$lorenz-ho.me, $$$=@>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+
+#include "Lv2UridMap.h"
+
+#ifdef LMMS_HAVE_LV2
+
+static LV2_URID staticMap(LV2_URID_Map_Handle handle, const char* uri)
+{
+	UridMap* map = static_cast<UridMap*>(handle);
+	return map->map(uri);
+}
+
+static const char* staticUnmap(LV2_URID_Unmap_Handle handle, LV2_URID urid)
+{
+	UridMap* map = static_cast<UridMap*>(handle);
+	return map->unmap(urid);
+}
+
+UridMap::UridMap()
+{
+	m_mapFeature.handle = static_cast<LV2_URID_Map_Handle>(this);
+	m_mapFeature.map = staticMap;
+	m_unmapFeature.handle = static_cast<LV2_URID_Unmap_Handle>(this);
+	m_unmapFeature.unmap = staticUnmap;
+}
+
+LV2_URID UridMap::map(const char *uri)
+{
+	LV2_URID result = 0u;
+	auto itr = m_map.find(uri);
```

Comment:
Do you mean lookup causes insertion?

Reply:
I mean concurrent modification and reading of standard library containers isn't safe. In the case of `std::unordered_map` in particular, insertion can cause a rehash. This could leave `itr` pointing to the wrong object, or even to freed memory. In the case of `std::vector`, insertion can cause a reallocation, which means that reading from it can access freed memory if it happens at the same time. To avoid this, make sure that `m_MapMutex` is locked both when reading from and when writing to `m_map` and `m_unMap`.

Reply:
Oh I get what you mean :smiley: . Makes sense.

Reply:
Should be fixed in 95d54fa